Brian branch

bhanford6 opened the pull request.

Commit:
implemented user hubs page

contains table of all the hubs a user is currently a part of
table contains hub name, creator, member count and rank

Commit:
user and hub profile updates

can now edit hub profile page
user lists now contain the user's picture and overall score
hub profile displays all user scores added together

no medal count implementations have been started yet
user lists are not yet ordered by highest score

vmaggioli reviewed (approved):
Do you make every CSS file by hand or do you have templates you use?  Those things are so long and complex

bhanford6 commented:
I've been making them by hand. For most of them I am able to use bits and pieces from other files though so It's easier to copy and paste. Really hard to read though. Probably should've been commenting them.

Commit:
rearrange backend to make it flatter

all of the data that is used for tracking scores of users and hubs is now its own json object
rankings service created for handling all scores data for both users and hubs

Commit:
implemented user and hub rankings

the rank of each user can now be viewed in profile
the rank of each hub can now be viewed in profile and user's list of hubs

Commit:
implemented leaderboards

leaderboards page shows top 100 users and hubs
lists are paginated into 10 item pages

UPDATED TO LATEST VERSION OF ANGULAR

Commit:
Touch-ups, notifs, Medal enums, init medal page

forgot to make multiple commits so there is a lot going on in this one

Notifications page is setup to show notifications for friend request, achievements and kicks, but achievements/medals and kicks are not fully implemented yet.
Enum files and values created to differentiate the different types of medals. Hoping that this will allow me to keep all general medal data in memory so as to not user that space in the database.
Empty medals page added and routed
Some minor aesthetic touchups throughout
Removed needless print statements througout.

Commit:
implementation of medals

Enumerated values for medals so they can be in memory instead of taking up space in the database
Objects for interaction of each type of medal
General medal object for controling types of medals
Scores and names given to each medal
Intervals of enum values given in a way that will allow for more medals to be added at a later time without overlapping values nor needed change of code implementation
Single page keeping tabbed interface for viewing all medals, only owned medals or a category of medals

Yet to implement for medals:
	Owned Medals
	Searching Medals
	Medal Profile Page
	Asigning score of medal to user and/or hub

Large commit caused by all of the default Medal data. Not much was changed to the logic of any of the code, just constantvalues created and interfaces for the enumerations of the medals so as to make sure that all of the medal data is kept in memory and not wasting space in database.

Commit:
rough one, a lot been goin on

everything regarding medals received a full 360
	instead of a strict number enumeration i took advantage of the javascript-json string enumeration ability so that each medal enumeration can be represented by two characters. this will reduce database size by 1 cahracter every time a medal is stored in the database. not huge, but enough of a difference to implement it.

	medals now have arrays and maps to make sure constant look up is available as well as ngFor capability (ngFor can only iterate over arrays)
		We don't care that it takes more lines of code and in-memory storage because we have no chance of every coming within 50% capacity of firebase hosting storage.

figured out how to use material pop-up windows and implemented that for each medal description instead of havinga fully separate component generated with its html and css like the rest of the components

each medal had its own object and enumeration so that it can be a type and a class with accessible members.

some fixes with friends tab of profile page because i didn't quite have the friend's score mapped to memory correctly

notifications page is not listening for changes to the number of friends that a member has and displaying a notification of a medal being received any time the user meets a milestone number of friends

just a notice that these enumerations of the medals make all coding related to the medals trivial. time-consuming and terribly boring to type out, but it requires no logic past making sure your type in the correct enumerated values. This makes it very easy to read and nice for autocompletion when typing into a decent IDE

there was probably more I missed. I'll try to do better with the committing.

Commit:
implemented medals and their notifications

all medals have been made and can be viewed from the medals tab
notifications have been made for pretty much all medals (not heavily tested)
popups made for showing more info about each medal
